src/commands/plugins.rs

```diff
@@ -0,0 +1,275 @@
+            get_spin_plugins_directory()?,
+            self.yes_to_all,
+            env!("VERGEN_BUILD_SEMVER"),
+        )
+        .install()
+        .await?;
+        Ok(())
+    }
+}
+
+/// Uninstalls specified plugin.
+#[derive(Parser, Debug)]
+pub struct Uninstall {
+    /// Name of Spin plugin.
+    pub name: String,
+}
+
+impl Uninstall {
+    pub async fn run(self) -> Result<()> {
+        PluginUninstaller::new(&self.name, get_spin_plugins_directory()?).run()?;
+        Ok(())
+    }
+}
+
+#[derive(Parser, Debug)]
+pub struct Upgrade {
+    /// Name of Spin plugin to upgrade.
+    #[clap(
+        name = "PLUGIN_NAME",
+        conflicts_with = "ALL",
+        required_unless_present_any = ["ALL"],
+    )]
+    pub name: Option<String>,
+
+    /// Upgrade all plugins.
+    #[clap(
+        short = 'a',
+        long = "all",
+        name = "ALL",
+        conflicts_with = "PLUGIN_NAME",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        takes_value = false
+    )]
+    pub all: bool,
+
+    /// Path to local plugin manifest.
+    #[clap(
+        name = "LOCAL_PLUGIN_MANIFEST",
+        short = 'f',
+        long = "file",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST"
+    )]
+    pub local_manifest_src: Option<PathBuf>,
+
+    /// Path to remote plugin manifest.
+    #[clap(
+        name = "REMOTE_PLUGIN_MANIFEST",
+        short = 'u',
+        long = "url",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST"
+    )]
+    pub remote_manifest_src: Option<Url>,
+
+    /// Skips prompt to accept the installation of the plugin[s].
+    #[clap(short = 'y', long = "yes", takes_value = false)]
+    pub yes_to_all: bool,
+
+    /// Specific version of a plugin to be install from the centralized plugins
+    /// repository.
+    #[clap(
+        long = "version",
+        short = 'v',
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        conflicts_with = "ALL",
+        requires("PLUGIN_NAME")
+    )]
+    pub version: Option<Version>,
+
+    /// Allow downgrading a plugin's version.
+    #[clap(short = 'd', long = "downgrade", takes_value = false)]
+    pub downgrade: bool,
+}
+
+impl Upgrade {
+    /// Upgrades one or all plugins by reinstalling the latest or a specified
+    /// version of a plugin. If downgrade is specified, first uninstalls the
+    /// plugin.
+    pub async fn run(self) -> Result<()> {
+        let plugins_dir = get_spin_plugins_directory()?;
+        let spin_version = env!("VERGEN_BUILD_SEMVER");
+        let manifest_dir = plugins_dir.join(PLUGIN_MANIFESTS_DIRECTORY_NAME);
+
+        // Check if no plugins are currently installed
+        if !manifest_dir.exists() {
+            println!("No currently installed plugins to update.");
+            return Ok(());
+        }
+
+        if self.all {
+            // Install the latest of all currently installed plugins
```

Comment:
Strongly suggest moving this block out to a `upgrade_all()` function.  The `else` is so far away, and there is so much indentation, it's hard for the reader to match it with the `if`.